src/coap_oscore.c

```diff
@@ -0,0 +1,2100 @@
+        goto error;
+      if (coap_request) {
+        /* Make as Inner option (unchanged) */
+        if (!coap_insert_option(plain_pdu,
+                                opt_iter.number,
+                                coap_opt_length(option),
+                                coap_opt_value(option)))
+          goto error;
+        osc_pdu->code = COAP_REQUEST_CODE_FETCH;
+      } else {
+        /* Make as Inner option but empty */
+        if (!coap_insert_option(plain_pdu, opt_iter.number, 0, NULL))
+          goto error;
+        osc_pdu->code = COAP_RESPONSE_CODE(205);
+      }
+      show_pdu = 1;
+      doing_observe = 1;
+      observe_value = coap_decode_var_bytes(coap_opt_value(option),
+                                            coap_opt_length(option));
+      break;
+    case COAP_OPTION_PROXY_URI:
+      /* Need to break down into the component parts RFC8613 4.1.3.3 */
+      memset(&uri, 0, sizeof(uri));
+      if (coap_split_proxy_uri(coap_opt_value(option),
+                               coap_opt_length(option),
+                               &uri) < 0) {
+        coap_log(LOG_WARNING,
+                 "Proxy URI '%.*s' not decodable\n",
+                 coap_opt_length(option),
+                 (const char *)coap_opt_value(option));
+        goto error;
+      }
+      /* Outer options */
+      if (!coap_insert_option(osc_pdu,
+                              COAP_OPTION_URI_HOST,
+                              uri.host.length,
+                              uri.host.s))
+        goto error;
+      if (uri.port != (coap_uri_scheme_is_secure(&uri) ? COAPS_DEFAULT_PORT
+                                                       : COAP_DEFAULT_PORT) &&
+          !coap_insert_option(osc_pdu,
+                              COAP_OPTION_URI_PORT,
+                              coap_encode_var_safe(option_value_buffer,
+                                                   sizeof(option_value_buffer),
+                                                   uri.port & 0xffff),
+                              option_value_buffer))
+        goto error;
+      if (uri.scheme >= COAP_URI_SCHEME_LAST ||
+          !coap_insert_option(osc_pdu,
+                              COAP_OPTION_PROXY_SCHEME,
+                              strlen(coap_uri_scheme[uri.scheme]),
+                              (const uint8_t *)coap_uri_scheme[uri.scheme]))
+        goto error;
+      /* Inner options */
+      if (uri.path.length) {
+        uint8_t *buf;
+        uint8_t *kbuf;
+        size_t buflen = uri.path.length + 1;
+        int res;
+
+        kbuf = buf = coap_malloc(uri.path.length + 1);
+        if (buf) {
+          res = coap_split_path(uri.path.s, uri.path.length, buf, &buflen);
+          while (res--) {
+            if (!coap_insert_option(plain_pdu,
+                                    COAP_OPTION_URI_PATH,
+                                    coap_opt_length(buf),
+                                    coap_opt_value(buf))) {
```

Comment:
I was trying with the URL `coap://[fd00::ff:fe00:2]/hello`. The URI path option was falsely set to `hello%00g`.

Reply:
Interesting.  This code is specifically for handling Proxy-URI in a case statement. What exactly is the command line invocation you are trying to use, as well as the contents of the file used for the OSCORE definitions as defined by the `-E` option.
````
$ examples/coap-client coap://[fd00::ff:fe00:2]/hello -v8 -E test/test_oscore.c_conf
````
reports
````
May 25 16:07:48.650 DEBG PDU presented by app
v:1 t:CON c:GET i:942c {01} [ Uri-Path:hello ]
````
and the server reports (assuming rfc8613_b_1_2 is set to false in the server configuration file)
````
v:1 t:CON c:GET i:0a34 {02000000000002} [ Uri-Path:hello ]
May 25 16:09:34.214 DEBG request for unknown resource 'hello', return 4.04
v:1 t:ACK c:4.04 i:0a34 {02000000000002} [ ] :: 'Not Found'
````

Reply:
Forget about it, the error was on my side.